runtime/src/impls.rs

```diff
@@ -123,6 +128,139 @@ impl Convert<(Weight, Fixed64), Fixed64> for FeeMultiplierUpdateHandler {
+{
+	/// Fill the gas meter
+	///
+	/// The process is as follows:
+	/// 1) Calculate the cost to fill the gas meter (gas price * gas limit)
+	/// 2a) Default case:
+	///    - User is paying in the native fee currency
+	///    - Deduct the 'fill meter cost' from the users balance and fill the gas meter
+	/// 2b) User has nominated to pay fees in another currency
+	///    - Calculate the 'fill gas cost' in terms of their nominated payment currency-
+	///      using the CENNZX spot exchange rate
+	///....- Check the user has liquid balance to pay the converted 'fill gas cost' and fill the gas meter
+	fn fill_gas(transactor: &T::AccountId, gas_limit: Gas) -> Result<GasMeter<T>, DispatchError> {
+		// Calculate the cost to fill the meter in the CENNZnet fee currency
+		let gas_price = Contracts::<T>::gas_price();
+		let fill_meter_cost = if gas_price.is_zero() {
+			// Gas is free in this configuration, fill the meter
+			return Ok(GasMeter::with_limit(gas_limit, gas_price));
+		} else {
+			gas_price
+				.checked_mul(&gas_limit.saturated_into())
+				.ok_or("Overflow during gas cost calculation")?
+		};
+
+		// Check if a fee exchange has been specified by the user
+		let fee_exchange: Option<FeeExchange<T::AssetId, T::Balance>> = storage::unhashed::get(&GAS_FEE_EXCHANGE_KEY);
+
+		if fee_exchange.is_none() {
+			// User will pay for gas in CENNZnet's native fee currency
+			let imbalance = T::Currency::withdraw(
+				transactor,
+				fill_meter_cost,
+				WithdrawReason::Fee.into(),
+				ExistenceRequirement::KeepAlive,
+			)?;
+			T::GasPayment::on_unbalanced(imbalance);
+			return Ok(GasMeter::with_limit(gas_limit, gas_price));
+		}
+
+		// User wants to pay fee in a nominated currency
+		let exchange_op = fee_exchange.unwrap();
+		let payment_asset = exchange_op.asset_id();
+
+		// Calculate the `fill_meter_cost` in terms of the user's nominated payment asset
+		let converted_fill_meter_cost = CennzxSpot::<T>::get_asset_to_core_output_price(
+			&payment_asset,
+			T::Balance::unique_saturated_from(fill_meter_cost.saturated_into()),
+			CennzxSpot::<T>::fee_rate(),
+		)?;
+
+		// Respect the user's max. fee preference
+		if converted_fill_meter_cost > exchange_op.max_payment() {
+			return Err("Fee cost exceeds max. payment limit".into());
+		}
+
+		// Does the user have enough funds to pay the `converted_fill_meter_cost` with `payment_asset`
+		// also taking into consideration any liquidity restrictions
+		GenericAsset::<T>::ensure_can_withdraw(
+			&payment_asset,
+			transactor,
+			converted_fill_meter_cost,
+			WithdrawReason::Fee.into(),
+			// Calculate the expected user balance after paying the `converted_fill_meter_cost`
+			// This value is required to ensure liquidity restrictions are upheld
+			GenericAsset::<T>::free_balance(&payment_asset, transactor)
+				.checked_sub(&converted_fill_meter_cost)
+				.ok_or("Insufficient liquidity to fill gas meter")?,
```

Comment:
Do we want to check the liquidity of exchange as well? (please ignore me if we don't need to)
```rust
let ex_liquidity = CennzxSpot::<T>::get_liquidity(
    &(CENTRAPAY_ASSET_ID, payment_asset),
    transactor,
);

if converted_fill_meter_cost > ex_liquidity {
    return Err("Fee cost exceeds exchange liquidity pool".into());
}
```

Reply:
Good point. @holygits what do you say?

Reply:
This is accounted for when `converted_fill_meter_cost` is calculated by CENNZ-X